packages/zoe/src/contracts/sellItems.js

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-use-before-define */
+// @ts-check
+
+import harden from '@agoric/harden';
+import { makeZoeHelpers, defaultAcceptanceMsg } from '../contractSupport';
+
+/** @typedef {import('../zoe').ContractFacet} ContractFacet */
+
+// zcf is the Zoe Contract Facet, i.e. the contract-facing API of Zoe
+export const makeContract = harden(
+  /** @param {ContractFacet} zcf */ zcf => {
+    const { assertKeywords, rejectOffer, checkHook } = makeZoeHelpers(zcf);
+    assertKeywords(harden(['Items', 'Money']));
+
+    const { terms } = zcf.getInstanceRecord();
+    const { pricePerItem } = terms;
+
+    let sellerOfferHandle;
+
+    const sellerOfferHook = offerHandle => {
+      sellerOfferHandle = offerHandle;
+      return defaultAcceptanceMsg;
+    };
+
+    const buyerOfferHook = buyerOfferHandle => {
+      const sellerAllocation = zcf.getCurrentAllocation(sellerOfferHandle);
+      const buyerAllocation = zcf.getCurrentAllocation(buyerOfferHandle);
+      const currentItemsForSale = sellerAllocation.Items;
+      const providedMoney = buyerAllocation.Money;
+
+      const { proposal } = zcf.getOffer(buyerOfferHandle);
+      const wantedItems = proposal.want.Items;
+      const numItemsWanted = wantedItems.extent.length;
+      const amountMaths = zcf.getAmountMaths(['Items', 'Money']);
+      const totalCostExtent = pricePerItem.extent * numItemsWanted;
+      const totalCost = amountMaths.Money.make(totalCostExtent);
+
+      // Check that the wanted items are still for sale.
+      if (!amountMaths.Items.isGTE(currentItemsForSale, wantedItems)) {
+        return rejectOffer(
+          buyerOfferHandle,
+          `Some of the wanted items were not available for sale`,
+        );
+      }
+
+      // Check that the money provided to pay for the items is greater than the totalCost.
+      if (!amountMaths.Money.isGTE(providedMoney, totalCost)) {
+        return rejectOffer(
+          buyerOfferHandle,
+          `More money (${pricePerItem.extent}) is required to buy these items`,
+        );
+      }
+
+      // Reallocate and complete the buyer offer.
+      const newSellerAllocation = {
```

Comment:
This is not for change now, but it seems wrong to have this much boilerplate for a transfer. It seems like ti could too easily get out of whack. It's protected by offer safety but evenso. Perhaps a helper method to take some allocations and a delta and produce the new allocations? (give me buyer and seller with x swapped for y).

Reply:
I made a new helper called `trade` which takes the delta in terms of gains for both offers and does the reallocation for you. I'm going to submit this change as a separate PR since it turned out to be a pretty large change (I used it in Autoswap and multipoolAutoswap).

Reply:
Also, thanks so much for noticing that this could be improved!